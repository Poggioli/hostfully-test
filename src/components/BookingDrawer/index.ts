export * from './BookingDrawer'
export * from './useBookingDrawer'