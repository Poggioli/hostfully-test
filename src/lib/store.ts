import { Booking } from "@/service/Booking";
import { Room } from "@/service/Room";
import { create } from "zustand";

type Store = {
  rooms: Room[];
  bookings: Booking[];
  findRoom: (id: string) => Room | undefined;
  setBookings: (bookings: Booking[]) => void;
  setRooms: (rooms: Room[]) => void;
  editBooking: (booking: Booking) => void;
  deleteBooking: (id: string) => void;
  addBooking: (booking: Booking) => void;
};

export const useStore = create<Store>((set, get) => ({
  rooms: [],
  bookings: [],
  setRooms: (rooms: Room[]) => set(() => ({ rooms })),
  findRoom: (id: string) => {
    const { rooms } = get()
    return rooms.find((room) => room.id === id)
  },
  setBookings: (bookings: Booking[]) => set(() => ({ bookings })),
  editBooking: (booking: Booking) =>
    set(({ bookings }) => ({
      bookings: bookings.map((b) => (booking.id === b.id ? booking : b)),
    })),
  deleteBooking: (id: string) =>
    set(({ bookings }) => ({
      bookings: bookings.filter((b) => b.id !== id),
    })),
  addBooking: (booking: Booking) =>
    set(({ bookings }) => ({
      bookings: [...bookings, booking],
    })),
}));
