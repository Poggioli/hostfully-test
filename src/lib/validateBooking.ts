const ONE_DAY_IN_MILLI = 86400000 as const;

export function isEndTimeGreatherThanStartTime(startTime: Date, endTime: Date): boolean {
  return endTime.getTime() > startTime.getTime();
}

export function isBookingAtLeastForOneDay(startTime: Date, endTime: Date): boolean {
    return endTime.getTime() - startTime.getTime() >= ONE_DAY_IN_MILLI;
}

export function isNotOverlapping(startTimeA: Date, endTimeA: Date, startTimeToBeBooked: Date, endTimeToBeBooked: Date): boolean {
  return (startTimeA.getTime() > startTimeToBeBooked.getTime() && startTimeA.getTime() > endTimeToBeBooked.getTime()) ||
      (startTimeToBeBooked.getTime() > endTimeA.getTime() && endTimeToBeBooked.getTime() > endTimeA.getTime())
}