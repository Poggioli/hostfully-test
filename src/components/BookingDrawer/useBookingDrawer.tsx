import { useStore } from "@/lib/store";
import { Room } from "@/service/Room";
import { useMemo } from "react";
import { useSearchParams } from "react-router-dom";

export const useBookingDrawer = () => {
  const { rooms } = useStore();
  const [searchParams, setSearchParams] = useSearchParams();
  const key = 'openBookingModal' as const

  const onClose = (): void => {
    const newSearchParams = new URLSearchParams(searchParams);
    newSearchParams.delete(key);
    setSearchParams(newSearchParams);
  }

  const onOpen = (id: string): void => {
    const newSearchParams = new URLSearchParams(searchParams)
    newSearchParams.set(key, id)
    setSearchParams(newSearchParams);
  }

  const room: Room | undefined = useMemo(() => {
    const id = searchParams.get(key) || '';
    return rooms.find((room) => room.id === id);
  }, [searchParams, rooms])

  const openBookingModal: boolean = useMemo(() => {
    return !!searchParams.get(key)
  }, [searchParams])

  return { onClose, onOpen, room, openBookingModal };
}