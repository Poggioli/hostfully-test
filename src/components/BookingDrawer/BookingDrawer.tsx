import { BookingForm } from "@/components/BookingForm";
import { Button } from "@/components/ui/button";
import { Drawer, DrawerContent, DrawerDescription, DrawerFooter, DrawerHeader, DrawerTitle } from "@/components/ui/drawer";
import { Separator } from "@/components/ui/separator";
import { FC } from "react";
import { useBookingDrawer } from "./useBookingDrawer";


type BookingDrawerErrorProps = {
  onClose: () => void;
}

const BookingDrawerError: FC<BookingDrawerErrorProps> = ({ onClose }) => {
  return (
    <>
      <DrawerHeader>
        <DrawerTitle className="text-center">Room not found</DrawerTitle>
        <DrawerDescription className="text-center">We could not found the room that you are looking for.</DrawerDescription>
      </DrawerHeader>
      <DrawerFooter>
        <Button onClick={onClose} variant="outline">Close</Button>
      </DrawerFooter>
    </>
  )
}

const BookingDrawer: FC = () => {
  const { onClose, room, openBookingModal } = useBookingDrawer();

  const onCloseInOnOpenChange = (value: boolean) => {
    if (!value) {
      onClose()
    }
  }

  return (
    <Drawer open={!!openBookingModal} onOpenChange={onCloseInOnOpenChange}>
      <DrawerContent>
        <div className="mx-auto max-w-[400px] max-h-[100vh] overflow-y-auto md:w-full md:max-w-[640px] md:overflow-hidden md:flex md:pt-8 md:items-center">
          {!room ? <BookingDrawerError onClose={onClose} /> : (
            <>
              <DrawerHeader className="md:max-w-[300px] md:w-full md:h-fit">
                <DrawerTitle className="text-left">{room.name}</DrawerTitle>
                <DrawerDescription className={"text-left"}>{room.description}</DrawerDescription>
              </DrawerHeader>
              <Separator className="max-w-[200px] mx-auto md:hidden" />
              <div className="p-4 md:w-full">
                <span className="text-sm mb-4 inline-block">
                  <span className="text-lg font-semibold">
                    ${room.pricePerDay}
                  </span>
                  {" "}night
                </span>
                <BookingForm room={room} />
              </div>
            </>
          )}
        </div>
      </DrawerContent>
    </Drawer>
  )
}

export { BookingDrawer };
