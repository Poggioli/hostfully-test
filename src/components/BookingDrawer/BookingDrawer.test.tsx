import { useStore } from "@/lib/store";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import { BookingDrawer } from "./BookingDrawer";

describe('BookingDrawer', () => {

  const renderComponent = (initialEntries?: string[]) => {
    const wrapper = ({ children }: { children: any }) => <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
    return render(<BookingDrawer />, { wrapper })
  }

  const mockStore = vi.fn().mockImplementation(useStore);

  mockStore.mockImplementation(() => ({
    rooms: [
      {
        id: "1",
        name: "Name 1",
        description: "Description 1",
        pricePerDay: 100,
        photos: []
      }
    ]
  }))

  it(`GIVEN a room
      WHEN room doesnt exists
      THEN should show the Not found message`, () => {
    const { baseElement } = renderComponent(['?openBookingModal=2']);
    expect(baseElement).toMatchSnapshot()
  })
})