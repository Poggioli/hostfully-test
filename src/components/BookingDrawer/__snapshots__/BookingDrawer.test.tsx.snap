// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookingDrawer > GIVEN a room
      WHEN room doesnt exists
      THEN should show the Not found message 1`] = `
<body
  style="position: fixed !important; top: 0px; left: 0px; right: 0px; height: auto; pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="opacity: 0; position: fixed; pointer-events: none; outline-color: none; outline-style: none; outline-width: initial;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="fixed inset-0 z-50 bg-black/80"
    data-aria-hidden="true"
    data-state="open"
    style="pointer-events: auto;"
    vaul-drawer-visible="true"
    vaul-overlay=""
    vaul-snap-points="false"
    vaul-snap-points-overlay="true"
  />
  <div
    aria-describedby="radix-:r2:"
    aria-labelledby="radix-:r1:"
    class="fixed inset-x-0 bottom-0 z-50 mt-24 flex h-auto flex-col rounded-t-[10px] border bg-background"
    data-state="open"
    id="radix-:r0:"
    role="dialog"
    style="pointer-events: auto;"
    tabindex="-1"
    vaul-drawer=""
    vaul-drawer-visible="true"
  >
    <div
      class="mx-auto mt-4 h-2 w-[100px] rounded-full bg-muted"
    />
    <div
      class="mx-auto max-w-[400px] max-h-[100vh] overflow-y-auto md:w-full md:max-w-[640px] md:overflow-hidden md:flex md:pt-8 md:items-center"
    >
      <div
        class="grid gap-1.5 p-4 text-center sm:text-left"
      >
        <h2
          class="text-lg font-semibold leading-none tracking-tight text-center"
          id="radix-:r1:"
        >
          Room not found
        </h2>
        <p
          class="text-sm text-muted-foreground text-center"
          id="radix-:r2:"
        >
          We could not found the room that you are looking for.
        </p>
      </div>
      <div
        class="mt-auto flex flex-col gap-2 p-4"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
        >
          Close
        </button>
      </div>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="opacity: 0; position: fixed; pointer-events: none; outline-color: none; outline-style: none; outline-width: initial;"
    tabindex="0"
  />
</body>
`;
